// nativescript/lib/common/constants.js
"use strict";
class ProvisionType {
}
ProvisionType.Development = "Development";
ProvisionType.AdHoc = "AdHoc";
ProvisionType.AppStore = "AppStore";
ProvisionType.Enterprise = "Enterprise";
exports.ProvisionType = ProvisionType;
exports.APP_RESOURCES_FOLDER_NAME = "App_Resources";
exports.ERROR_NO_DEVICES = "Cannot find connected devices. Reconnect any connected devices, verify that your system recognizes them, and run this command again.";
exports.UNREACHABLE_STATUS = "Unreachable";
exports.CONNECTED_STATUS = "Connected";
class LiveSyncConstants {
}
LiveSyncConstants.VERSION_2 = 2;
LiveSyncConstants.VERSION_3 = 3;
LiveSyncConstants.GUID = "12590FAA-5EDD-4B12-856D-F52A0A1599F2";
LiveSyncConstants.DEVICE_TMP_DIR_FORMAT_V2 = `/data/local/tmp/${LiveSyncConstants.GUID}/%s`;
LiveSyncConstants.ANDROID_FILES_PATH = `files/${LiveSyncConstants.GUID}`;
LiveSyncConstants.DEVICE_TMP_DIR_FORMAT_V3 = `/mnt/sdcard/Android/data/%s/${LiveSyncConstants.ANDROID_FILES_PATH}`;
LiveSyncConstants.CHECK_LIVESYNC_INTENT_NAME = "com.telerik.IsLiveSyncSupported";
LiveSyncConstants.IOS_PROJECT_PATH = "/Documents/AppBuilder/LiveSync";
exports.LiveSyncConstants = LiveSyncConstants;
exports.TARGET_FRAMEWORK_IDENTIFIERS = {
    Cordova: "Cordova",
    NativeScript: "NativeScript"
};
class Configurations {
}
Configurations.Debug = "Debug";
Configurations.Release = "Release";
exports.Configurations = Configurations;
exports.NODE_MODULES_DIR_NAME = "node_modules";
exports.TNS_CORE_MODULES = "tns-core-modules";
class FileExtensions {
}
FileExtensions.TYPESCRIPT_DEFINITION_FILE = ".d.ts";
FileExtensions.TYPESCRIPT_FILE = ".ts";
exports.FileExtensions = FileExtensions;
