// nativescript/lib/common/appbuilder/mobile-platforms-capabilities.js
"use strict";
class MobilePlatformsCapabilities {
    constructor($errors) {
        this.$errors = $errors;
    }
    getPlatformNames() {
        return _.keys(this.getAllCapabilities());
    }
    getAllCapabilities() {
        this.platformCapabilities = this.platformCapabilities || {
            iOS: {
                wirelessDeploy: true,
                cableDeploy: true,
                companion: true,
                hostPlatformsForDeploy: ["win32", "darwin"]
            },
            Android: {
                wirelessDeploy: true,
                cableDeploy: true,
                companion: true,
                hostPlatformsForDeploy: ["win32", "darwin", "linux"]
            },
            WP8: {
                wirelessDeploy: true,
                cableDeploy: false,
                companion: true,
                hostPlatformsForDeploy: ["win32"]
            }
        };
        return this.platformCapabilities;
    }
}
exports.MobilePlatformsCapabilities = MobilePlatformsCapabilities;
$injector.register("mobilePlatformsCapabilities", MobilePlatformsCapabilities);
