// nativescript/lib/commands/emulate.js
"use strict";
class EmulateCommandBase {
    constructor($platformService) {
        this.$platformService = $platformService;
    }
    executeCore(args) {
        return this.$platformService.emulatePlatform(args[0]);
    }
}
exports.EmulateCommandBase = EmulateCommandBase;
class EmulateIosCommand extends EmulateCommandBase {
    constructor($platformService, $platformsData) {
        super($platformService);
        this.$platformsData = $platformsData;
        this.allowedParameters = [];
    }
    execute(args) {
        return this.executeCore([this.$platformsData.availablePlatforms.iOS]);
    }
}
exports.EmulateIosCommand = EmulateIosCommand;
$injector.registerCommand("emulate|ios", EmulateIosCommand);
class EmulateAndroidCommand extends EmulateCommandBase {
    constructor($platformService, $platformsData) {
        super($platformService);
        this.$platformsData = $platformsData;
        this.allowedParameters = [];
    }
    execute(args) {
        return this.executeCore([this.$platformsData.availablePlatforms.Android]);
    }
}
exports.EmulateAndroidCommand = EmulateAndroidCommand;
$injector.registerCommand("emulate|android", EmulateAndroidCommand);
