// nativescript/lib/common/commands/device/list-applications.js
"use strict";
const os_1 = require("os");
const util = require("util");
class ListApplicationsCommand {
    constructor($devicesService, $logger, $options) {
        this.$devicesService = $devicesService;
        this.$logger = $logger;
        this.$options = $options;
        this.allowedParameters = [];
    }
    execute(args) {
        return (() => {
            this.$devicesService.initialize({ deviceId: this.$options.device, skipInferPlatform: true }).wait();
            let output = [];
            let action = (device) => {
                return (() => {
                    let applications = device.applicationManager.getInstalledApplications().wait();
                    output.push(util.format("%s=====Installed applications on device with UDID '%s' are:", os_1.EOL, device.deviceInfo.identifier));
                    _.each(applications, (applicationId) => output.push(applicationId));
                }).future()();
            };
            this.$devicesService.execute(action).wait();
            this.$logger.out(output.join(os_1.EOL));
        }).future()();
    }
}
exports.ListApplicationsCommand = ListApplicationsCommand;
$injector.registerCommand("device|list-applications", ListApplicationsCommand);
