// nativescript/lib/common/services/plugins/npmjs-plugins-source.js
"use strict";
const parse5 = require("parse5");
const plugins_source_base_1 = require("./plugins-source-base");
class NpmjsPluginsSource extends plugins_source_base_1.PluginsSourceBase {
    constructor($progressIndicator, $logger, $httpClient, $childProcess, $hostInfo, $errors) {
        super($progressIndicator, $logger);
        this.$httpClient = $httpClient;
        this.$childProcess = $childProcess;
        this.$hostInfo = $hostInfo;
        this.$errors = $errors;
        this._pages = [];
    }
    get progressIndicatorMessage() {
        return "Searching for plugins in http://npmjs.org.";
    }
    getPlugins(page, count) {
        return (() => {
            let loadedPlugins = this._pages[page];
            if (loadedPlugins) {
                return loadedPlugins;
            }
            let result = this.getPluginsFromNpmjs(this._keywords, page).wait();
            this._pages[page] = result;
            this.plugins = this.plugins.concat(result);
            return result;
        }).future()();
    }
    getAllPlugins() {
        return (() => {
            let getAllPluginsFuture = this.getAllPluginsCore();
            this.$logger.printInfoMessageOnSameLine("Getting all results, please wait.");
            this.$progressIndicator.showProgressIndicator(getAllPluginsFuture, 2000).wait();
            return getAllPluginsFuture.get();
        }).future()();
    }
    initializeCore(projectDir, keywords) {
        return (() => {
            this._keywords = keywords;
            this.plugins = this.getPluginsFromNpmjs(keywords, 1).wait();
        }).future()();
    }
    getAllPluginsCore() {
        return (() => {
            let result = [];
            let currentPluginsFound = [];
            let page = 1;
            do {
                currentPluginsFound = this.getPluginsFromNpmjs(this._keywords, page++).wait();
                if (currentPluginsFound && currentPluginsFound.length) {
                    result = result.concat(currentPluginsFound);
                }
            } while (currentPluginsFound && currentPluginsFound.length);
            return result;
        }).future()();
    }
    getPluginsFromNpmjs(keywords, page) {
        return (() => {
            let pluginName = encodeURIComponent(keywords.join(" "));
            let url = `${NpmjsPluginsSource.NPMJS_ADDRESS}/search?q=${pluginName}&page=${page}`;
            try {
                let responseBody = this.$httpClient.httpRequest(url).wait().body;
                let document = parse5.parse(responseBody);
                let html = _.find(document.childNodes, (node) => node.nodeName === "html");
                let resultsContainer = this.findNodeByClass(html, "search-results");
                if (!resultsContainer || !resultsContainer.childNodes) {
                    return null;
                }
                let resultsElements = _.filter(resultsContainer.childNodes, (node) => node.nodeName === "li");
                return _.map(resultsElements, (node) => this.getPluginInfo(node));
            }
            catch (err) {
                this.$logger.trace(`Error while getting information for ${keywords} from http://npmjs.org - ${err}`);
                return null;
            }
        }).future()();
    }
    getPluginInfo(node) {
        let name = this.getTextFromElementWithClass(node, "name");
        let version = this.getTextFromElementWithClass(node, "version");
        let description = this.getTextFromElementWithClass(node, "description");
        let author = this.getTextFromElementWithClass(node, "author");
        return {
            name,
            version,
            description,
            author
        };
    }
    findNodeByClass(parent, className) {
        if (!parent.childNodes || parent.childNodes.length === 0) {
            return null;
        }
        for (let i = 0; i < parent.childNodes.length; i++) {
            let node = parent.childNodes[i];
            if (_.some(node.attrs, (attr) => attr.name === "class" && attr.value === className)) {
                return node;
            }
            else {
                let result = this.findNodeByClass(node, className);
                if (result) {
                    return result;
                }
            }
        }
    }
    getTextFromElementWithClass(node, className) {
        let element = this.findNodeByClass(node, className);
        if (element && element.childNodes) {
            let textElement = _.find(element.childNodes, (child) => child.nodeName === "#text");
            if (textElement) {
                return textElement.value;
            }
        }
        return null;
    }
}
NpmjsPluginsSource.NPMJS_ADDRESS = "http://npmjs.org";
exports.NpmjsPluginsSource = NpmjsPluginsSource;
