// nativescript/lib/common/appbuilder/providers/device-app-data-provider.js
"use strict";
const path = require("path");
const util = require("util");
const appbuilder_device_app_data_base_1 = require("../mobile/appbuilder-device-app-data-base");
const appbuilder_companion_device_app_data_base_1 = require("../mobile/appbuilder-companion-device-app-data-base");
const constants_1 = require("../../constants");
const Future = require("fibers/future");
class AndroidAppIdentifier extends appbuilder_device_app_data_base_1.AppBuilderDeviceAppDataBase {
    constructor(_appIdentifier, device, platform, $deployHelper, $devicePlatformsConstants, $errors) {
        super(_appIdentifier, device, platform, $deployHelper, $devicePlatformsConstants);
        this.$errors = $errors;
        this._deviceProjectRootPath = null;
    }
    get deviceProjectRootPath() {
        if (!this._deviceProjectRootPath) {
            let deviceTmpDirFormat = "";
            let version = this.getLiveSyncVersion().wait();
            if (version === 2) {
                deviceTmpDirFormat = constants_1.LiveSyncConstants.DEVICE_TMP_DIR_FORMAT_V2;
            }
            else if (version === 3) {
                deviceTmpDirFormat = constants_1.LiveSyncConstants.DEVICE_TMP_DIR_FORMAT_V3;
            }
            else {
                this.$errors.failWithoutHelp(`Unsupported LiveSync version: ${version}`);
            }
            this._deviceProjectRootPath = this.getDeviceProjectRootPath(util.format(deviceTmpDirFormat, this.appIdentifier));
        }
        return this._deviceProjectRootPath;
    }
    encodeLiveSyncHostUri(hostUri) {
        return hostUri;
    }
    isLiveSyncSupported() {
        return (() => {
            return super.isLiveSyncSupported().wait() && this.getLiveSyncVersion().wait() !== 0;
        }).future()();
    }
    getLiveSyncVersion() {
        return (() => {
            if (!this._liveSyncVersion) {
                this._liveSyncVersion = this.device.adb.sendBroadcastToDevice(constants_1.LiveSyncConstants.CHECK_LIVESYNC_INTENT_NAME, { "app-id": this.appIdentifier }).wait();
            }
            return this._liveSyncVersion;
        }).future()();
    }
}
exports.AndroidAppIdentifier = AndroidAppIdentifier;
class AndroidCompanionAppIdentifier extends appbuilder_companion_device_app_data_base_1.AppBuilderCompanionDeviceAppDataBase {
    constructor(device, platform, $deployHelper, $devicePlatformsConstants, $companionAppsService) {
        super($companionAppsService.getCompanionAppIdentifier(constants_1.TARGET_FRAMEWORK_IDENTIFIERS.Cordova, platform), device, platform, $deployHelper, $devicePlatformsConstants);
        this.$companionAppsService = $companionAppsService;
    }
    get deviceProjectRootPath() {
        return this.getDeviceProjectRootPath(util.format(constants_1.LiveSyncConstants.DEVICE_TMP_DIR_FORMAT_V3, this.appIdentifier));
    }
    get liveSyncFormat() {
        return "icenium://%s?token=%s&appId=%s&configuration=%s";
    }
    getCompanionAppName() {
        return "companion app";
    }
}
exports.AndroidCompanionAppIdentifier = AndroidCompanionAppIdentifier;
class AndroidNativeScriptCompanionAppIdentifier extends appbuilder_companion_device_app_data_base_1.AppBuilderCompanionDeviceAppDataBase {
    constructor(device, platform, $deployHelper, $devicePlatformsConstants, $companionAppsService) {
        super($companionAppsService.getCompanionAppIdentifier(constants_1.TARGET_FRAMEWORK_IDENTIFIERS.NativeScript, platform), device, platform, $deployHelper, $devicePlatformsConstants);
        this.$companionAppsService = $companionAppsService;
    }
    get deviceProjectRootPath() {
        return util.format(constants_1.LiveSyncConstants.DEVICE_TMP_DIR_FORMAT_V3, this.appIdentifier);
    }
    get liveSyncFormat() {
        return "nativescript://%s?token=%s&appId=%s&configuration=%s";
    }
    getCompanionAppName() {
        return "NativeScript companion app";
    }
}
exports.AndroidNativeScriptCompanionAppIdentifier = AndroidNativeScriptCompanionAppIdentifier;
class IOSAppIdentifier extends appbuilder_device_app_data_base_1.AppBuilderDeviceAppDataBase {
    constructor(_appIdentifier, device, platform, $deployHelper, $devicePlatformsConstants, $iOSSimResolver) {
        super(_appIdentifier, device, platform, $deployHelper, $devicePlatformsConstants);
        this.$iOSSimResolver = $iOSSimResolver;
        this._deviceProjectRootPath = null;
    }
    get deviceProjectRootPath() {
        if (!this._deviceProjectRootPath) {
            if (this.device.isEmulator) {
                let applicationPath = this.$iOSSimResolver.iOSSim.getApplicationPath(this.device.deviceInfo.identifier, this.appIdentifier);
                this._deviceProjectRootPath = path.join(applicationPath, "www");
            }
            else {
                this._deviceProjectRootPath = constants_1.LiveSyncConstants.IOS_PROJECT_PATH;
            }
        }
        return this._deviceProjectRootPath;
    }
    getLiveSyncNotSupportedError() {
        return `You can't LiveSync on device with id ${this.device.deviceInfo.identifier}! Deploy the app with LiveSync enabled and wait for the initial start up before LiveSyncing.`;
    }
}
exports.IOSAppIdentifier = IOSAppIdentifier;
class IOSNativeScriptAppIdentifier extends appbuilder_device_app_data_base_1.AppBuilderDeviceAppDataBase {
    constructor(_appIdentifier, device, platform, $deployHelper, $devicePlatformsConstants, $iOSSimResolver) {
        super(_appIdentifier, device, platform, $deployHelper, $devicePlatformsConstants);
        this.$iOSSimResolver = $iOSSimResolver;
        this._deviceProjectRootPath = null;
    }
    get deviceProjectRootPath() {
        if (!this._deviceProjectRootPath) {
            if (this.device.isEmulator) {
                let applicationPath = this.$iOSSimResolver.iOSSim.getApplicationPath(this.device.deviceInfo.identifier, this.appIdentifier);
                this._deviceProjectRootPath = applicationPath;
            }
            else {
                this._deviceProjectRootPath = constants_1.LiveSyncConstants.IOS_PROJECT_PATH;
            }
        }
        return this._deviceProjectRootPath;
    }
}
exports.IOSNativeScriptAppIdentifier = IOSNativeScriptAppIdentifier;
class IOSCompanionAppIdentifier extends appbuilder_companion_device_app_data_base_1.AppBuilderCompanionDeviceAppDataBase {
    constructor(device, platform, $deployHelper, $devicePlatformsConstants, $companionAppsService) {
        super($companionAppsService.getCompanionAppIdentifier(constants_1.TARGET_FRAMEWORK_IDENTIFIERS.Cordova, platform), device, platform, $deployHelper, $devicePlatformsConstants);
        this.$companionAppsService = $companionAppsService;
    }
    get deviceProjectRootPath() {
        return constants_1.LiveSyncConstants.IOS_PROJECT_PATH;
    }
    get liveSyncFormat() {
        return "icenium://%s?LiveSyncToken=%s&appId=%s&configuration=%s";
    }
    getCompanionAppName() {
        return "companion app";
    }
}
exports.IOSCompanionAppIdentifier = IOSCompanionAppIdentifier;
class IOSNativeScriptCompanionAppIdentifier extends appbuilder_companion_device_app_data_base_1.AppBuilderCompanionDeviceAppDataBase {
    constructor(device, platform, $deployHelper, $devicePlatformsConstants, $companionAppsService) {
        super($companionAppsService.getCompanionAppIdentifier(constants_1.TARGET_FRAMEWORK_IDENTIFIERS.NativeScript, platform), device, platform, $deployHelper, $devicePlatformsConstants);
        this.$companionAppsService = $companionAppsService;
    }
    get deviceProjectRootPath() {
        return constants_1.LiveSyncConstants.IOS_PROJECT_PATH;
    }
    get liveSyncFormat() {
        return "nativescript://%s?LiveSyncToken=%s&appId=%s&configuration=%s";
    }
    getCompanionAppName() {
        return "NativeScript companion app";
    }
}
exports.IOSNativeScriptCompanionAppIdentifier = IOSNativeScriptCompanionAppIdentifier;
class WP8CompanionAppIdentifier extends appbuilder_companion_device_app_data_base_1.AppBuilderCompanionDeviceAppDataBase {
    constructor(device, $deployHelper, $devicePlatformsConstants, platform, $companionAppsService) {
        super($companionAppsService.getCompanionAppIdentifier(constants_1.TARGET_FRAMEWORK_IDENTIFIERS.Cordova, platform), device, platform, $deployHelper, $devicePlatformsConstants);
        this.platform = platform;
        this.$companionAppsService = $companionAppsService;
    }
    get deviceProjectRootPath() {
        return "";
    }
    get liveSyncFormat() {
        return "%s/Mist/MobilePackage/redirect?token=%s&appId=%s&configuration=%s";
    }
    encodeLiveSyncHostUri(hostUri) {
        return hostUri;
    }
    isLiveSyncSupported() {
        return Future.fromResult(true);
    }
    getLiveSyncNotSupportedError() {
        return "";
    }
    getCompanionAppName() {
        return "companion app";
    }
}
exports.WP8CompanionAppIdentifier = WP8CompanionAppIdentifier;
class DeviceAppDataProvider {
    constructor($project) {
        this.$project = $project;
    }
    createFactoryRules() {
        let rules = {
            Cordova: {
                Android: {
                    vanilla: AndroidAppIdentifier,
                    companion: AndroidCompanionAppIdentifier
                },
                iOS: {
                    vanilla: IOSAppIdentifier,
                    companion: IOSCompanionAppIdentifier
                },
                WP8: {
                    vanilla: "",
                    companion: WP8CompanionAppIdentifier
                }
            },
            NativeScript: {
                Android: {
                    vanilla: AndroidAppIdentifier,
                    companion: AndroidNativeScriptCompanionAppIdentifier
                },
                iOS: {
                    vanilla: IOSNativeScriptAppIdentifier,
                    companion: IOSNativeScriptCompanionAppIdentifier
                }
            }
        };
        return rules[this.$project.projectData.Framework];
    }
}
exports.DeviceAppDataProvider = DeviceAppDataProvider;
$injector.register("deviceAppDataProvider", DeviceAppDataProvider);
