// nativescript/lib/common/services/plugins/print-plugins-service.js
"use strict";
const helpers_1 = require("../../helpers");
class PrintPluginsService {
    constructor($errors, $logger, $prompter) {
        this.$errors = $errors;
        this.$logger = $logger;
        this.$prompter = $prompter;
        this._page = 1;
    }
    printPlugins(pluginsSource, options) {
        return (() => {
            if (!pluginsSource.hasPlugins()) {
                this.$logger.warn("No plugins found.");
                return;
            }
            let count = options.count || PrintPluginsService.COUNT_OF_PLUGINS_TO_DISPLAY;
            if (!helpers_1.isInteractive() || options.showAllPlugins) {
                let allPlugins = pluginsSource.getAllPlugins().wait();
                this.displayTableWithPlugins(allPlugins);
                return;
            }
            let pluginsToDisplay = pluginsSource.getPlugins(this._page++, count).wait();
            let shouldDisplayMorePlugins = true;
            this.$logger.out("Available plugins:");
            do {
                this.displayTableWithPlugins(pluginsToDisplay);
                if (pluginsToDisplay.length < count) {
                    return;
                }
                shouldDisplayMorePlugins = this.$prompter.confirm("Load more plugins?").wait();
                pluginsToDisplay = pluginsSource.getPlugins(this._page++, count).wait();
                if (!pluginsToDisplay || pluginsToDisplay.length < 1) {
                    return;
                }
            } while (shouldDisplayMorePlugins);
        }).future()();
    }
    displayTableWithPlugins(plugins) {
        let data = [];
        data = this.createTableCells(plugins);
        let table = this.createPluginsTable(data);
        this.$logger.out(table.toString());
    }
    createPluginsTable(data) {
        let headers = ["Plugin", "Version", "Author", "Description"];
        let table = helpers_1.createTable(headers, data);
        return table;
    }
    createTableCells(plugins) {
        return _.map(plugins, (plugin) => [plugin.name, plugin.version, plugin.author || "", plugin.description || ""]);
    }
}
PrintPluginsService.COUNT_OF_PLUGINS_TO_DISPLAY = 10;
exports.PrintPluginsService = PrintPluginsService;
$injector.register("printPluginsService", PrintPluginsService);
