// nativescript/lib/common/mobile/android/android-debug-bridge.js
"use strict";
class AndroidDebugBridge {
    constructor($childProcess, $errors, $logger, $staticConfig, $androidDebugBridgeResultHandler) {
        this.$childProcess = $childProcess;
        this.$errors = $errors;
        this.$logger = $logger;
        this.$staticConfig = $staticConfig;
        this.$androidDebugBridgeResultHandler = $androidDebugBridgeResultHandler;
    }
    executeCommand(args, options) {
        return (() => {
            let event = "close";
            let command = this.composeCommand(args).wait();
            let treatErrorsAsWarnings = false;
            let childProcessOptions = undefined;
            if (options) {
                event = options.fromEvent || event;
                treatErrorsAsWarnings = options.treatErrorsAsWarnings;
                childProcessOptions = options.childProcessOptions;
                if (options.returnChildProcess) {
                    return this.$childProcess.spawn(command.command, command.args);
                }
            }
            let result = this.$childProcess.spawnFromEvent(command.command, command.args, event, childProcessOptions, { throwError: false }).wait();
            let errors = this.$androidDebugBridgeResultHandler.checkForErrors(result);
            if (errors && errors.length > 0) {
                this.$androidDebugBridgeResultHandler.handleErrors(errors, treatErrorsAsWarnings);
            }
            return (result.stdout === undefined || result.stdout === null) ? result : result.stdout;
        }).future()();
    }
    composeCommand(params, identifier) {
        return (() => {
            let command = this.$staticConfig.getAdbFilePath().wait();
            let deviceIdentifier = [];
            if (identifier) {
                deviceIdentifier = ["-s", `${identifier}`];
            }
            let args = deviceIdentifier.concat(params);
            return { command, args };
        }).future()();
    }
}
exports.AndroidDebugBridge = AndroidDebugBridge;
$injector.register("adb", AndroidDebugBridge);
