// nativescript/lib/common/mobile/ios/simulator/ios-simulator-log-provider.js
"use strict";
const fiberBootstrap = require("../../../fiber-bootstrap");
class IOSSimulatorLogProvider {
    constructor($iOSSimResolver, $deviceLogProvider, $devicePlatformsConstants, $processService) {
        this.$iOSSimResolver = $iOSSimResolver;
        this.$deviceLogProvider = $deviceLogProvider;
        this.$devicePlatformsConstants = $devicePlatformsConstants;
        this.$processService = $processService;
    }
    startLogProcess(deviceIdentifier) {
        if (!this.isStarted) {
            let deviceLogChildProcess = this.$iOSSimResolver.iOSSim.getDeviceLogProcess(deviceIdentifier);
            let action = (data) => {
                fiberBootstrap.run(() => this.$deviceLogProvider.logData(data.toString(), this.$devicePlatformsConstants.iOS, deviceIdentifier));
            };
            if (deviceLogChildProcess.stdout) {
                deviceLogChildProcess.stdout.on("data", action);
            }
            if (deviceLogChildProcess.stderr) {
                deviceLogChildProcess.stderr.on("data", action);
            }
            this.$processService.attachToProcessExitSignals(this, deviceLogChildProcess.kill);
            this.isStarted = true;
        }
    }
}
exports.IOSSimulatorLogProvider = IOSSimulatorLogProvider;
$injector.register("iOSSimulatorLogProvider", IOSSimulatorLogProvider);
